Deferred/Deferred.swift

```diff
@@ -57,52 +149,28 @@ public struct Deferred<Value> {
     /// :param: value The resolved value of the deferred.
     /// :param: assertIfFilled If `false`, race checking is disabled.
     public func fill(value: Value, assertIfFilled: Bool = true, file: StaticString = __FILE__, line: UInt = __LINE__) {
-        let (filledValue, blocks) = protected.withWriteLock { data -> (Value, [UponBlock]) in
-            if assertIfFilled {
-                precondition(data.protected == nil, "Cannot fill an already-filled Deferred", file: file, line: line)
-                data.protected = value
-            } else if data.protected == nil {
-                data.protected = value
-            }
-            let blocks = data.uponBlocks
-            data.uponBlocks.removeAll(keepCapacity: false)
-            return (data.protected!, blocks)
-        }
-        for (queue, block) in blocks {
-            dispatch_async(queue) { block(filledValue) }
+        let succeeded = storage.fill(value, onFill: markFilled)
+        switch (succeeded, assertIfFilled) {
+        case (false, true):
+            preconditionFailure("Cannot fill an already-filled Deferred", file: file, line: line)
+        case (_, _):
+            break
         }
     }
     
     /**
-    Checks for and returns a determined value.
-    
-    :returns: The determined value, if already filled, or `nil`.
-    */
-    public func peek() -> Value? {
-        return protected.withReadLock { $0.protected }
-    }
-    
-    /**
     Call some function once the value is determined.
     
     If the value is already determined, the function will be submitted to the
     queue immediately. An `upon` call is always executed asynchronously.
     
     :param: queue A dispatch queue for executing the given function on.
-    :param: function A function that uses the determined value.
+    :param: body A function that uses the determined value.
     */
-    public func upon(queue: dispatch_queue_t = DeferredDefaultQueue, function: Value -> ()) {
-        let maybeValue: Value? = protected.withWriteLock{ data in
-            if data.protected == nil {
-                data.uponBlocks.append( (queue, function) )
-            }
-            return data.protected
-        }
-        if let value = maybeValue {
-            dispatch_async(queue) { function(value) }
-        }
+    public func upon(queue: dispatch_queue_t = genericQueue, body: Value -> ()) {
+        _ = upon(queue, options: DISPATCH_BLOCK_INHERIT_QOS_CLASS, body: body)
```

Comment:
It might have a `WARN_UNUSED_VALUE` jobby attached, or it's just to let the reader know that the value is intentionally being discarded, similar to doing `(void)printf("foo\n");` to go "yes, this returns something, but I am intentionally ignoring that fact".